File: app/src/main/java/cz/ackee/sample/login/LoginPresenter.java
package cz.ackee.sample.login;

import cz.ackee.sample.interactor.ApiInteractorImpl;
import cz.ackee.sample.model.LoginResponse;
import io.reactivex.android.schedulers.AndroidSchedulers;

/**
 * Presenter for login screen
 * Created by David Bilik[david.bilik@ackee.cz] on {05/08/16}
 **/
public class LoginPresenter {
    public static final String TAG = LoginPresenter.class.getName();
    private final ApiInteractorImpl apiInteractor;
    private ILoginView view;

    public LoginPresenter() {
        this.apiInteractor = new ApiInteractorImpl();
    }

    public void login(String name, String password) {
        this.apiInteractor.login(name, password)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onLoggedIn, this::onErrorHappened);
    }

    public void onViewAttached(ILoginView view) {
        this.view = view;
    }

    public void onViewDetached() {
        this.view = null;
    }

    private void onLoggedIn(LoginResponse loginResponse) {
        if(view != null) {
            view.openDetail();
        }
    }

    private void onErrorHappened(Throwable throwable) {
        throwable.printStackTrace();
    }

}

File: app/src/main/java/cz/ackee/sample/login/LoginPresenter.kt
package cz.ackee.sample.login

import cz.ackee.sample.App
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

/**
 * Presenter for login screen
 */
class LoginPresenter {

    private val apiInteractor = App.diContainer.apiInteractor
    private var view: ILoginView? = null

    fun login(name: String, password: String) {
        this.apiInteractor.login(name, password)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ this.onLoggedIn() }, { this.onErrorHappened(it) })
    }

    fun onViewAttached(view: ILoginView) {
        this.view = view
    }

    fun onViewDetached() {
        this.view = null
    }

    private fun onLoggedIn() {
        view?.openDetail()
    }

    private fun onErrorHappened(throwable: Throwable) {
        throwable.printStackTrace()
    }
}
